### Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | actual cat | actual dog 
|:------------  |-----------:|-----------:|
| predicted cat |         34 |          7 |
| predicted dog |         13 |         46 |


- In the context of this problem, what is a false positive?
- In the context of this problem, what is a false negative?
- How would you describe this model?

In [1]:
import pandas as pd

In [ ]:
# In the context of this problem, what is a false positive?
# False alarm. Predicted cat and it was an actual dog. Value == 7

In [ ]:
# In the context of this problem, what is a false negative?
# Miss. Predicted dog and was actually a cat. Value == 13

In [6]:
# How would you describe this model?
TP = 34
TN = 46
FP = 7
FN = 13
accuracy = (TP + TN)/(TP + TN + FP + FN) * 100
accuracy
print(f"The model is {accuracy}% accurate.")

The model is 80.0% accurate.


### You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [ ]:
# Which evaluation metric would be appropriate here?
# The team wants to investigate the defects. 

In [30]:
r_duck = pd.read_csv('untidy_data/c3.csv')
r_duck.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [31]:
# Value counts to see how many defects and non defects there are
r_duck.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [52]:
# Baseline shows there are more no defects than defects.
# Basline is to predict no defects, so we should use specificity to calculate TN out of
# all Actual Negatives
actual_v_m1 = pd.crosstab(r_duck.model1, r_duck.actual)
actual_v_m1

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [53]:
# Model_1 accuracy and specificity
TP = 8
TN = 182
FP = 2
FN = 8
accuracy = (TP + TN)/(TP + TN + FP + FN) * 100
accuracy
print(f"The model is {accuracy}% accurate.")
# % of predicting TN out of all actual Negatives
specificity = TN/(TN + FP) * 100
print(f"The models specificity is {round(specificity,2)}%.")

The model is 95.0% accurate.
The models specificity is 98.91%.


In [54]:
actual_v_m2 = pd.crosstab(r_duck.model2, r_duck.actual)
actual_v_m2

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [56]:
# Model_2 accuracy and specificty
TP = 9
TN = 103
FP = 81
FN = 7
accuracy = (TP + TN)/(TP + TN + FP + FN) * 100
accuracy
print(f"The model is {round(accuracy,2)}% accurate.")
# % of predicting TN out of all actual Negatives
specificity = TN/(TN + FP) * 100
print(f"The models specificity is {round(specificity,2)}%.")

The model is 56.0% accurate.
The models specificity is 55.98%.


In [57]:
actual_v_m3 = pd.crosstab(r_duck.model3, r_duck.actual)
actual_v_m3

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [58]:
# Model_3 accuracy and specificity
TP = 13
TN = 98
FP = 86
FN = 3
accuracy = (TP + TN)/(TP + TN + FP + FN) * 100
accuracy
print(f"The model is {round(accuracy,2)}% accurate.")
# % of predicting TN out of all actual Negatives
specificity = TN/(TN + FP) * 100
print(f"The models specificity is {round(specificity,2)}%.")

The model is 55.5% accurate.
The models specificity is 53.26%.


### Model 1 shows the highes specificity that is predicting the TN out of all actual negatives at 98.91%.

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [59]:
# They want us to predict defects but don't want to give out a vacation if there is no
# defect. We would pick the precision metric or % positive predictions that are correct
# model_1
TP = 8
TN = 182
FP = 2
FN = 8

precision = TP/(TP + FP) * 100
print(f"The models precision is {round(precision,2)}%.")

The models precision is 80.0%.


In [60]:
# model_2
TP = 9
TN = 103
FP = 81
FN = 7
precision = TP/(TP + FP) * 100
print(f"The models precision is {round(precision,2)}%.")

The models precision is 10.0%.


In [61]:
# model_3
TP = 13
TN = 98
FP = 86
FN = 3
precision = TP/(TP + FP) * 100
print(f"The models precision is {round(precision,2)}%.")

The models precision is 13.13%.


### Based on the 3 models. Model_1 show the highest precision or the model the highest positive predictive value at 80.00%

### 3. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

- At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

- Several models have already been developed with the data, and you can find their results here.

- Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [48]:
paws = pd.read_csv('untidy_data/gives_you_paws.csv')
paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [49]:
# Show more dogs than cats. Baseline is that they will give you a dog pic
paws.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

### A.In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [62]:
# Model_1
a_v_m1 = pd.crosstab(paws.model1, paws.actual)
a_v_m1

actual   cat   dog
model1            
cat     1423   640
dog      323  2614

In [63]:
# In terms of accuracy, how do the various models compare to the baseline model? Are any
# of the models better than the baseline?
TP = 1423
TN = 2614
FP = 640
FN = 323

accuracy = (TP + TN)/(TP + TN + FP + FN) * 100
accuracy
print(f"The model is {round(accuracy,2)}% accurate.")

The model is 80.74% accurate.


In [64]:
a_v_m1 = pd.crosstab(paws.model2, paws.actual)
a_v_m1

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [65]:
!git add 

On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   classification_exercises.ipynb

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	EvaluatingModelPerformance.pdf
	Python_Seaborn_Cheat_Sheet.pdf
	explore_practice.ipynb
	iris_df.csv
	model_evaluation.ipynb
	untidy_data/

no changes added to commit (use "git add" and/or "git commit -a")
